lib/spack/spack/test/concretize.py

```diff
@@ -458,19 +463,71 @@ def test_concretize_two_virtuals_with_dual_provider_and_a_conflict(self):
     @pytest.mark.only_clingo(
         "Optional compiler propagation isn't deprecated for original concretizer"
     )
+    @pytest.mark.parametrize("unify", [True, False, "when_possible"])
+    def test_concretize_environment_propagated_disabled_variant(
+        self, unify, tmpdir, mutable_mock_env_path
+    ):
+        path = tmpdir.join("spack.yaml")
+
+        with tmpdir.as_cwd():
+            with open(str(path), "w") as f:
+                f.write(
+                    """\
+spack:
+  specs:
+    - hypre ~~shared ^openblas
+"""
+                )
+
+        SpackCommand("env")("create", "test", str(path))
+
+        test = spack.environment.read("test")
+        test.unify = unify
+        test.concretize()
+
+        for spec in test.specs_by_hash.values():
+            for dep in spec.dependencies():
+                if dep.name == "openblas":
+                    assert dep.satisfies("~shared")
+                    assert dep.satisfies("^zlib ~shared")
+
+    @pytest.mark.skipif(
+        os.environ.get("SPACK_TEST_SOLVER") == "original",
+        reason="Optional compiler propagation isn't deprecated for original concretizer",
+    )
     def test_concretize_propagate_disabled_variant(self):
         """Test a package variant value was passed from its parent."""
         spec = Spec("hypre~~shared ^openblas")
         spec.concretize()
 
         assert spec.satisfies("^openblas~shared")
+        assert spec.satisfies("^zlib~shared")
 
+    @pytest.mark.skipif(
+        os.environ.get("SPACK_TEST_SOLVER") == "original",
+        reason="Optional compiler propagation isn't deprecated for original concretizer",
+    )
     def test_concretize_propagated_variant_is_not_passed_to_dependent(self):
         """Test a package variant value was passed from its parent."""
         spec = Spec("hypre~~shared ^openblas+shared")
         spec.concretize()
 
         assert spec.satisfies("^openblas+shared")
+        assert spec.satisfies("^zlib~shared")
+
+    @pytest.mark.skipif(
+        os.environ.get("SPACK_TEST_SOLVER") == "original",
+        reason="Optional compiler propagation isn't deprecated for original concretizer",
+    )
+    def test_concretize_propagate_specified_variant(self):
+        """Test that only the specified variant is propagated to the dependencies"""
+        spec = Spec("splice-b++bar")
+        spec.concretize()
+
+        assert spec.satisfies("^splice-a+bar")
+        assert spec.satisfies("^splice-z+bar")
+        assert not spec.satisfies("^splice-a+foo")
+        assert not spec.satisfies("^splice-z+foo")
```

Comment:
Something like:
```suggestion
        assert spec.satisfies("+bar") and spec.satisfies("^splice-a+bar")
        assert spec.satisfies("+bar") and spec.satisfies("^splice-z+bar")
        assert spec.satisfies("+foo") and not spec.satisfies("^splice-a+foo")
        assert spec.satisfies("+foo") and not spec.satisfies("^splice-z+foo")
```
would, in my opinion, make clear that one is propagated and the other not.